Revisit Users context, cover UI with e2e tests and introduce first AuditLog features

AndrewDryga opened the pull request:
1. `auto_create_users` default value is removed. We want to avoid situations when admins integrate OIDC/SAML providers and don't expect anyone that has access to it to automatically gain access to VPN, which is especially critical for providers like Google Workspace, where all employees typically have access.
2. OpenID library was completely rewritten and a new version is integrated. It will allow async tests and better scales for the cloud version of the panel.
3. `Mox` was removed, we don't test modules by overriding them to prevent breaking changes that tests can't capture.
4. Deps are reordered and unused ones are removed.
5. Browser/e2e tests are added to ensure we won't break UI features in the future, allowing for front-end refactoring.
6. Users context was overhauled for better code clarity.

TODO:
- [ ] Move authentication concerns from the web to the domain app
- [ ] Add authentication AuditLog which will be a stepping stone for a complete audit log for all changes made in the Panel.
- [ ] Track count of recent failed login attempts for a user and throttle rapid requests
- [x] Finish covering admin UI with e2e tests
- [x] Configure CI to run e2e tests

AndrewDryga committed:
Refactor half of Users context

AndrewDryga committed:
Remove dead code

AndrewDryga committed:
Finish refactoring of User context

AndrewDryga committed:
Stop credo complaining about single with clauses

AndrewDryga committed:
Make dialyzer happy

AndrewDryga committed:
Offset the IP lock by hash on CIDR network to prevent races in tests

AndrewDryga committed:
Remove unused deps and reorder remaining ones

AndrewDryga committed:
Add Wallaby helpers and first browser test

AndrewDryga committed:
Add sandbox bindings to LiveView and Channels

AndrewDryga committed:
Added tests for OIDC login

AndrewDryga committed:
Add more tests

AndrewDryga committed:
Do not auto-create OIDC users by default

We don't want to take this security decision for our users, because some providers (like Google Workspace) will allow anyone with corporate email to log-in into potentially security-critical services by default, so this responsibility is for the user to take from now on.

AndrewDryga committed:
Start the endpoint as last element of supervision tree

We don't want Panel to accept any HTTP connections before all other components are up and running, otherwise:
1. a rapid request after the panel boot will return an error and even force supervisor tree to restart;
2. most of deployment targets use HTTP or PORT readiness probes and we don't want them to assume that container is ready before it's actually ready to accept requests, otherwise it's added to the load balancer too early.

AndrewDryga committed:
Do not output logs while tests are running

It adds a lot of noise to the output. The new configuration will output captured logs only when test has failed.

jamilbk commented on apps/fz_http/assets/js/live_view.js:
```diff
@@ -20,8 +20,7 @@ const channelToken = document
 const notificationChannel =
   userSocket.channel("notification:session", {
-    token: channelToken,
-    user_agent: window.navigator.userAgent
```
Hmm I think removing this line might break the `Active Sessions` table?

<img width="1616" alt="Screenshot 2023-01-09 at 5 14 33 PM" src="https://user-images.githubusercontent.com/167144/211439738-3eada60a-f450-4fb4-b74c-0b95dfbb3c9c.png">

AndrewDryga replied:
@jamilbk it should not, I'm getting the `user_agent` in `UserSocket` instead. Phoenix has built-in `:user_agent` option in `Phoenix.Socket`'s `:connect_info` so there is no sense to duplicate it by sending it from JavaScript as a parameter.

Anyway, I'll have a test for it :).

jamilbk replied:
Awesome :-)

AndrewDryga commented on mix.lock:
```diff
@@ -28,15 +28,12 @@
   "file_size": {:hex, :file_size, "3.0.1", "ad447a69442a82fc701765a73992d7b1110136fa0d4a9d3190ea685d60034dcd", [:mix], [{:decimal, ">= 1.0.0 and < 3.0.0", [hex: :decimal, repo: "hexpm", optional: true]}, {:number, "~> 1.0", [hex: :number, repo: "hexpm", optional: false]}], "hexpm", "64dd665bc37920480c249785788265f5d42e98830d757c6a477b3246703b8e20"},
   "file_system": {:hex, :file_system, "0.2.10", "fb082005a9cd1711c05b5248710f8826b02d7d1784e7c3451f9c1231d4fc162d", [:mix], [], "hexpm", "41195edbfb562a593726eda3b3e8b103a309b733ad25f3d642ba49696bf715dc"},
-  "floki": {:hex, :floki, "0.34.0", "002d0cc194b48794d74711731db004fafeb328fe676976f160685262d43706a8", [:mix], [], "hexpm", "9c3a9f43f40dde00332a589bd9d389b90c1f518aef500364d00636acc5ebc99c"},
-  "gen_smtp": {:hex, :gen_smtp, "1.2.0", "9cfc75c72a8821588b9b9fe947ae5ab2aed95a052b81237e0928633a13276fd3", [:rebar3], [{:ranch, ">= 1.8.0", [hex: :ranch, repo: "hexpm", optional: false]}], "hexpm", "5ee0375680bca8f20c4d85f58c2894441443a743355430ff33a783fe03296779"},
```
I think maybe I should not remove this, is it used? @jamilbk

jamilbk replied:
Hmmm I can't say I remember where those are used, if at all. If nothing is depending on them, we could remove.

jamilbk replied:
Ah, I think these are used in the `test` env.

AndrewDryga replied:
I think it was included because the Swoosh SMTP adapter requires it to send emails, so If we remove it then the SMTP adapter will stop working. I'll bring it back.

AndrewDryga committed:
Remove default value for auto_create_users, integrate new OIDC lib and add more e2e tests

AndrewDryga committed:
Add chromedriver to CI

AndrewDryga committed:
Fix post-rebase issues

AndrewDryga committed:
Make credo happy

AndrewDryga committed:
Reorganize shared deps

AndrewDryga committed:
Restructure formatter configs

AndrewDryga committed:
Bump deps

AndrewDryga committed:
Remove duplicate dep

AndrewDryga committed:
Change application supervisor names

AndrewDryga committed:
Persist screenshots for failed cases

AndrewDryga committed:
Add JUnit formatter for better GitHub Actions integration

AndrewDryga committed:
Bump OpenIDConnect version

AndrewDryga committed:
Save screenshots on failure

AndrewDryga committed:
Add separate steps to compile deps and app to keep track of their execution time

AndrewDryga committed:
Build and cache assets during test execution

AndrewDryga committed:
Add Vault to CI

AndrewDryga committed:
Introduce wait assertion

AndrewDryga committed:
Fix module imports

AndrewDryga committed:
Simplify tests with helpers and increase wait timeout

AndrewDryga committed:
Trim trailing whitespaces

AndrewDryga committed:
Add OIDC test when local auth is disabled

AndrewDryga committed:
Remove TODO

AndrewDryga committed:
Add tests for MFA removal

AndrewDryga committed:
Add tests for unpiv user device management options

AndrewDryga committed:
Improve some wait helpers and expose ENV variable for wait time

AndrewDryga committed:
Started adding admin UI tests

AndrewDryga committed:
Make credo happy

AndrewDryga committed:
Fix failed test

AndrewDryga committed:
Add root path for unauthenticated user

Fixes: https://github.com/firezone/firezone/pull/1267#issuecomment-1381311162

AndrewDryga committed:
Eat this, credo!

AndrewDryga committed:
Make test less flaky and fix a bug

AndrewDryga committed:
Add tests for API tokens

AndrewDryga committed:
Remove current_password field to unify forms

This field did not add any security because it was only used in form when admin user updated himself, but he can also go to his profile in Users list and update the same fields without confirming existing password, so there is no real security need for it.

AndrewDryga committed:
Do not show confirmation alert when user clicks disabled user delete button

AndrewDryga committed:
Added tests for user profile

AndrewDryga committed:
Added tests for security settings

AndrewDryga committed:
Add test for managing allow rules

AndrewDryga committed:
Add user management tests

AndrewDryga committed:
Disable acceptance cases by default

AndrewDryga committed:
Split acceptance cases in separate job and run multiple partitions of it

AndrewDryga committed:
Fix partitioning params

AndrewDryga committed:
Add newline to end of CI config

AndrewDryga committed:
Rerun failed acceptance tests automatically

jamilbk commented on .github/workflows/test.yml:
```diff
@@ -54,11 +54,147 @@ jobs:
+                   --no-deps-check \
+              || mix test --failed
+      - name: Save Screenshots
+        if: success() || failure()
```
`always()` could be used here?

jamilbk commented on apps/fz_http/lib/fz_http/authorization.ex:
```diff
@@ -0,0 +1,11 @@
+defmodule FzHttp.Authorization do
+  # def get_subject(%FzHttp.Users.User{} = user, user_ip, user_agent) do
```
Just double-checking that leaving these commented out was intentional

AndrewDryga replied:
I'll remove it, it will be part of the next PR.

jamilbk reviewed (dismissed):
Looks great! Just a couple minor comments

AndrewDryga committed:
Always save screenshots

AndrewDryga committed:
Lock junit formatter version

AndrewDryga committed:
Remove files that will go into next PR
